src/service/resource.js

```diff
@@ -654,32 +651,31 @@ export class Resource {
   }
 
   /**
+   * @param {number|boolean} viewport
    * @return {!Promise} resolves when underlying element is built and within the
-   *    range specified by implementation's renderOutsideViewport
+   *    viewport range given.
    */
-  whenWithinRenderOutsideViewport() {
-    if (!this.isLayoutPending()) {
+  whenWithinViewport(viewport) {
+    dev().assert(viewport !== false);
+    if (!this.isLayoutPending() || this.hasOwner() || viewport === true ||
+        this.withinViewportMultiplier(viewport)) {
       return Promise.resolve();
     }
-    if (this.renderOutsideViewportPromise_) {
-      return this.renderOutsideViewportPromise_;
-    }
-    const deferred = new Deferred();
-    this.renderOutsideViewportResolve_ = deferred.resolve;
-    return this.renderOutsideViewportPromise_ = deferred.promise;
+    const viewportNum = dev().assertNumber(viewport);
+    this.withViewportDeferreds_[viewportNum] =
+        this.withViewportDeferreds_[viewportNum] || new Deferred();
+    return this.withViewportDeferreds_[viewportNum].promise;
   }
 
-  /**
-   * @private resolves render outside viewport promise if one was created via
-   *    whenWithinRenderOutsideViewport.
-   */
-  resolveRenderOutsideViewport_() {
-    if (!this.renderOutsideViewportResolve_) {
-      return;
-    }
-    this.renderOutsideViewportResolve_();
-    this.renderOutsideViewportPromise_ = null;
-    this.renderOutsideViewportResolve_ = null;
+  /** @private resolves promises populated via whenWithinViewport. */
+  resolveWithinViewports_() {
```

Comment:
name is vague, maybe something more descriptive like `resolveDeferredsWhenWithinViewports` ?